File: asg4/List.c
//Antony Robbins
//androbbi@ucsc.edu
//Assignment 2
#include<stdio.h>
#include<stdlib.h>
#include "List.h"

// private NodeObj type
typedef struct NodeObj{
   int data;
   struct NodeObj* next;
   struct NodeObj* prev;
} NodeObj;

// private Node type
typedef NodeObj* Node;

// private ListObj type
typedef struct ListObj{
   Node front;
   Node back;
   Node cursor;
   int length;
   int index;
} ListObj;

// Constructors-Destructors ---------------------------------------------------
// newNode()
// Returns reference to new Node object. Initializes next and data fields.
// Private.
Node newNode(int data){
   Node N = malloc(sizeof(NodeObj));
   N->data = data;
   N->next = NULL;
   N->prev = NULL;
   return(N);
}

// newList()
// Returns reference to new empty List object.
List newList(void){
   List L;
   L = malloc(sizeof(ListObj));
   L->front = L->back = L->cursor = NULL;
   L->length = 0;
   L->index = -1;
   return(L);
}


// freeNode()
// Frees heap memory pointed to by *pN, sets *pN to NULL.
// Private.
void freeNode(Node* pN){
   if( pN!=NULL && *pN!=NULL ){
      free(*pN);
      *pN = NULL;
   }
}


// freeList()
// Frees all heap memory associated with List *pQ, and sets *pQ to NULL.S
void freeList(List* pL){
   if(pL!=NULL && *pL!=NULL) { 
	  if( length(*pL) > 0)
         clear(*pL);
      free(*pL);
      *pL = NULL;
   }
}
 
// Access functions -----------------------------------------------------------

// getFront()
// Returns the value at the front of L.
// Pre: !isEmpty(L)
int front(List L){
   if( L==NULL ){
      printf("List Error: calling front() on NULL List reference\n");
      exit(1);
   }
   if( length(L) <= 0 ){
      printf("List Error: calling front() on an empty List\n");
      exit(1);
   }
   return(L->front->data);
}

// getLength()
// Returns the length of Q.
int length(List L){
   if( L==NULL ){
      printf("List Error: calling length() on NULL List reference\n");
      exit(1);
   }
   return(L->length);
}
/*
// isEmpty()
// Returns true (1) if L is empty, otherwise returns false (0)
int isEmpty(List L){
   if( L==NULL ){
      printf("List Error: calling isEmpty() on NULL List reference\n");
      exit(1);
   }
   return(L->length==0);
}*/

// Manipulation procedures ----------------------------------------------------

// append()
// Places new data element at the end of L
void append(List L, int data)
{
   Node tmp = newNode(data);
   if( L==NULL ){
      printf("List Error: calling append() on NULL List reference\n");
      exit(1);
   }
   if( length(L) <= 0 ) { L->front = L->back = tmp; }
   else { tmp->prev = L->back; L->back->next = tmp; L->back = tmp; }
   L->length++;
}

// Returns the index of the cursor element in this list, or 
// returns -1 if the cursor element is undefined.
int getIndex(List L){
   if(L->cursor != NULL){
	  return L->index;
   }
   else
	  return -1;
}
   
// Returns back element in this List. Pre: length()>0
int back(List L){
   if(length(L) <0){
      printf("List Error: calling back() on NULL List reference\n");
      exit(1);
   }
   return L->back->data; 
}
   
// Returns cursor element in this list. 
// Pre: length()>0, getIndex()>=0
int getElement(List L) {
   if(length(L) <= 0 || getIndex(L) < 0){
      printf("List Error: calling getElement() on NULL List reference, or an index with -1\n");
      exit(1);
   }
   return L->cursor->data;
}

// Inserts new element before front element in this List. Index does increment
void prepend(List L, int data) {
   Node tmp = newNode(data);
   if(length(L) == 0){
      L->front = tmp;
      L->back= L->front;
   }
   else{
	  tmp->next = L->front;
      L->front->prev = tmp;
      L->front = tmp;
      if(L->cursor!= NULL) {L->index++;}
   }
	L->length++;
} 

// Pre: !this.isEmpty()
// Deletes the front element in this List. Pre: length()>0
void deleteFront(List L){
   Node tmp = NULL;
   if( L==NULL ){
      printf("List Error: calling deleteFront on NULL List reference\n");
      exit(1);
   }
   if( length(L) <= 0 ){
      printf("List Error: calling deleteFront on an empty List\n");
      exit(1);
   }
   if( length(L) == 1) {
      tmp = L->front;
      L->front = L->back = NULL;
	  freeNode(&tmp);
   }
   if( length(L)>1 ) {
      tmp = L->front;
      L->front = L->front->next;
	  L->front->prev = NULL;
	  freeNode(&tmp);
   }

   if (L->index == 0) {
      L->cursor = NULL;
	  L->index = -1;
   }
   if(L->cursor!=NULL) {L->index--;}
   L->length--;
}
   
// If 0<=i<=length()-1, moves the cursor to the element 
// at index i, otherwise the cursor becomes undefined.   
void moveTo(List L, int i) {
   Node tmp = NULL;
   if (i >= 0 && i <= length(L)-1) {
      tmp = L->front;
      for(int j = 0; j < i; j++) {
	     tmp = tmp->next;
	  }
      L->cursor = tmp;
      L->index = i;
   }
   else{
	  L->cursor = NULL;
	  L->index = -1;
   }		 
}
   
// If 0<getIndex()<=length()-1, moves the cursor one step toward the
// front of the list. If getIndex()==0, cursor becomes undefined. 
// If getIndex()==-1, cursor remains undefined. This operation is
// equivalent to moveTo(getIndex()-1).
void movePrev(List L) {
   if(0 < getIndex(L) && getIndex(L) <= length(L) -1) {
	  moveTo(L, getIndex(L)-1);
   }	 
   else if (getIndex(L) == 0 || getIndex(L) == -1){
      L->cursor = NULL;
	  L->index = -1;
   }
}

// If 0<=getIndex()<length()-1, moves the cursor one step toward the
// back of the list. If getIndex()==length()-1, cursor becomes
// undefined. If index==-1, cursor remains undefined. This
// operation is equivalent to moveTo(getIndex()+1).
void moveNext(List L) {
   if(0 <= getIndex(L) && getIndex(L) <= length(L) -1) {
	  moveTo(L, getIndex(L)+1);		 
   }	 
   else if (getIndex(L) == length(L)-1 || getIndex(L) == -1){
	  L->cursor = NULL;
	  L->index = -1;
   }
}  
   
// Inserts new element before cursor element in this
// List. Pre: length()>0, getIndex()>=0
void insertBefore(List L, int data) {
   if(length(L)>0 && getIndex(L) >= 0) {
	  Node in = newNode(data);
	  Node tmp = NULL;
      tmp = L->cursor->prev;
      tmp->next = in;
      L->cursor->prev = in;
      in->prev = tmp;
      in->next = L->cursor;
      L->index++;
      L->length++; 
   }
}
   
// Inserts new element after cursor element in this 
// List. Pre: length()>0, getIndex()>=0
void insertAfter(List L, int data) {
   if(length(L)>0 && getIndex(L) >= 0) {
	  Node in = newNode(data);
	  Node tmp = NULL;
      tmp = L->cursor->next;
      tmp->prev = in;
      L->cursor->next = in;
      in->prev = L->cursor;
      in->next = tmp;
      L->length++;
   }
}
   
// Re-sets this List to the empty state.
void clear(List L) {
   Node tmp = NULL;
   while(L->front != NULL){
	  tmp = L->front;
      L->front = L->front->next;
      freeNode(&tmp);
   }
   freeNode(&tmp);
   L->length = 0;
   L->index = -1;
}
   
// Deletes the back element in this List. Pre: length()>0
void deleteBack(List L) {
   if(length(L) > 0){
      Node tmp = NULL;
	  tmp = L->back;
	  L->back = L->back->prev;
	  L->length--;
	  freeNode(&tmp);
	  if(L->index == length(L) -1) {L->cursor = NULL; L->index = -1;}
   }
}

// Deletes cursor element in this List. Cursor is undefined after this 
// operation. Pre: length()>0, getIndex()>=0
void delete(List L) {
   if(length(L) > 0 && getIndex(L)>=0){
      Node tmp = NULL;
	  Node left = NULL;
	  Node right = NULL;
      tmp = L->cursor;
      left = tmp->prev;
      right = tmp->next;
      left->next = right;
      right->prev = left;
      L->cursor = NULL;
	  freeNode(&tmp);
      L->index = -1;
	  L->length--;
   }
   //else 
	 // throw new RuntimeException("List Error: delete() when cursor is null");
}

// Other Functions ------------------------------------------------------------

// printList()
// Prints data elements in L on a single line to stdout.
void printList(FILE* out, List L){
   Node curr = NULL;

   if( L==NULL ){
      printf("List Error: calling printList() on NULL List reference\n");
      exit(1);
   }
   for(curr = L->front; curr != NULL; curr = curr->next){
      fprintf(out, "%d ", curr->data);
   }
   //printf("\n");
}

// equals()
// returns true (1) if A is identical to B, false (0) otherwise
int equals(List A, List B){
   int flag = 1;
   Node N = NULL;
   Node M = NULL;

   if( A==NULL || B==NULL ){
      printf("List Error: calling equals() on NULL List reference\n");
      exit(1);
   }
   N = A->front;
   M = B->front;
   if( A->length != B->length ) { return 0; }
   while( flag && N!=NULL){
      flag = (N->data==M->data);
      N = N->next;
      M = M->next;
   }
   return flag;
}

// copy(): returns a new List identical to this one.
// Returns a new list representing the same integer sequence as this
// list. The cursor in the new list is undefined, regardless of the
// state of the cursor in this List. This List is unchanged.
   List copyList(List L){
      List M = newList();
      Node N = L->front;

      while( N!=NULL ){
         append(M, N->data);
         N = N->next;
      }
      return M;
   }

File: asg3/List.java
//Antony Robbins
//androbbi@ucsc.edu
//Assignment 3
class List{
   private class Node{
      // Fields
      Object data;
      Node next;
	  Node prev;
      // Constructor
      Node(Object data) {
	     this.data = data;
	     next = null;
		 prev = null;
	  }
	  
      // toString:  Overides Object's toString method.
      public String toString() { return String.valueOf(data); }
   } 

      // Fields
   private Node front;
   private Node back;
   private Node cursor;
   private int length;
   private int index;
  
   // Constructor
   List() {
      front = back = cursor = null; length = 0; index = -1;
   }
 
// Other methods 
// Access Functions //////////////////////////////////////////////////////////////////

   // Returns front element in this list. Pre: length()>0
   Object front(){
      if(length()>0){
         return front.data; 
      }
	  else
	     throw new RuntimeException("List Error: front() called on empty List");
   }

   // length(): returns length of this queue
   int length() { return length; }
   
  
   // isEmpty(): returns true if this is an empty queue, false otherwise
   //boolean isEmpty() { return length == 0; }
 
   
   // Returns the index of the cursor element in this list, or 
   // returns -1 if the cursor element is undefined.
   int getIndex(){
      if(cursor != null){
	     return index;
	  }
	  else
	     return -1;
   }
   
   // Returns back element in this List. Pre: length()>0
   Object back(){
      if(length()>0){
         return back.data; 
      }
	  else
	     throw new RuntimeException("List Error: back() called on empty List");
   }
   
   // Returns cursor element in this list. 
   // Pre: length()>0, getIndex()>=0
   Object getElement() {
      if(length() > 0 && getIndex() >= 0){
	     return cursor.data;
	  }
	  else
	     throw new RuntimeException("List Error: getElement is called on an empty List, or there is no cursor");
   }
   
   // Returns true if this List and L are the same Object
   // sequence. The cursor is ignored in both lists.
   // equals(): returns true if this List is identical to  L, false otherwise.
   public boolean equals(Object x){
	  if ( !(x instanceof Node) ) return false;
      if (x == this) return true;
      Node N = (Node)x;
      
      Node M = (Node)front;
      return N.data.equals(M.data);
   }


// Manipulation Procedures ///////////////////////////////////////////////////////////

   // Inserts new element after back element in this List.
   void append(Object data){
      Node tmp = new Node(data);
      if( length() <= 0 ) {
	     front = tmp;
		 back = front; 
	  }
      else {
	     tmp.prev = back;
	     back.next = tmp;
		 back = tmp;
	  }
	  length++;
   }
   
   // Inserts new element before front element in this List. Index does increment
   void prepend(Object data) {
      Node tmp = new Node(data);
      if(length() == 0){
         front = tmp;
		 back = front;
      }
      else{
	     tmp.next = front;
         front.prev = tmp;
		 front = tmp;
		 if(cursor!= null) {index++;}
      }
	   length++;
   } 

   // Pre: !this.isEmpty()
   // Deletes the front element in this List. Pre: length()>0
   void deleteFront(){
      if(length() <= 0){
         throw new RuntimeException("List Error: deleteFront() called on empty List");
      }
      if(this.length>1) {
	     Node tmp = front;
	     front = front.next;
		 front.prev = null;
		 tmp = null;
	  }
	  else if (index == 0) {cursor = null; index = -1;}
      else {front = back = null;}
	  if(cursor != null) { index--;}
      length--;
   }
   
   // If 0<=i<=length()-1, moves the cursor to the element 
   // at index i, otherwise the cursor becomes undefined.   
   void moveTo(int i) {
      if (i >= 0 && i <= length()-1) {
         Node tmp = front;
		 for(int j = 0; j < i; j++) {
		    tmp = tmp.next;
		 }
		 cursor = tmp;
		 index = i;
	  }
	  else{
	     cursor = null;
	     index = -1;
      }		 
   }
   
   // If 0<getIndex()<=length()-1, moves the cursor one step toward the
   // front of the list. If getIndex()==0, cursor becomes undefined. 
   // If getIndex()==-1, cursor remains undefined. This operation is
   // equivalent to moveTo(getIndex()-1).
   void movePrev() {
      if(0 < getIndex() && getIndex() <= length() -1) {
	     moveTo(getIndex()-1);
	  }	 
	  else if (getIndex() == 0 || getIndex() == -1){
	     cursor = null;
	     index = -1;
	  }
   }
   
   // If 0<=getIndex()<length()-1, moves the cursor one step toward the
   // back of the list. If getIndex()==length()-1, cursor becomes
   // undefined. If index==-1, cursor remains undefined. This
   // operation is equivalent to moveTo(getIndex()+1).
   void moveNext() {
      if(0 <= getIndex() && getIndex() <= length() -1) {
	     moveTo(getIndex()+1);		 
	  }	 
	  else if (getIndex() == length()-1 || getIndex() == -1){
	     cursor = null;
	     index = -1;
	  }
   }  
   
   // Inserts new element before cursor element in this
   // List. Pre: length()>0, getIndex()>=0
   void insertBefore(Object data) {
      if(length()>0 && getIndex() >= 0) {
	     Node in = new Node(data);
		 Node temp = cursor.prev;
		 temp.next = in;
		 cursor.prev = in;
		 in.prev = temp;
		 in.next = cursor;
		 index++;
		 length++;
		 
	   }
   }
   
   // Inserts new element after cursor element in this 
   // List. Pre: length()>0, getIndex()>=0
   void insertAfter(Object data) {
      if(length()>0 && getIndex() >= 0) {
	     Node in = new Node(data);
		 Node temp = cursor.next;
		 temp.prev = in;
		 cursor.next = in;
		 in.prev = cursor;
		 in.next = temp;
		 length++;
      }
   }
   
   // Re-sets this List to the empty state.
   void clear() {
      for(int i = 0; i < length; i++){
	     Node tmp = front;
		 front = front.next;
		 tmp = null;
	  }
	  front = back = cursor = null;
	  length = 0;
	  index = -1;
   }
   
   // Deletes the back element in this List. Pre: length()>0
   void deleteBack() {
      if(length() > 0){
	     Node tmp = back;
	     back = back.prev;
		 tmp = null;
	     length--;
	     if(index == length() -1) {cursor = null; index = -1;}
      }
	  else
	     throw new RuntimeException("List Error: deleteBack() called on empty List");
   }

   // Deletes cursor element in this List. Cursor is undefined after this 
   // operation. Pre: length()>0, getIndex()>=0
   void delete() {
      if(length() > 0 && getIndex()>=0){
		 Node tmp = cursor;
		 Node left = tmp.prev;
		 Node right = tmp.next;
         left.next = right;
		 right.prev = left;
		 tmp = null;
		 cursor = null;
         index = -1;
	     length--;
      }
      else 
	     throw new RuntimeException("List Error: delete() when cursor is null");
   }

   // Other Functions ///////////////////////////////////////////////////////////////////

   // toString():  overides Object's toString() method.
   // Overides Object's toString method. Creates a string 
   // consisting of a space separated sequence of Objects
   // with front on the left and back on the right. The
   // cursor is ignored.
   public String toString(){
      String str = "";
      for(Node N=front; N!=null; N=N.next){
         str += N.toString() + " ";
      }
      return str;
   }

/*
// copy(): returns a new Queue identical to this one.
// Returns a new list representing the same Object sequence as this
// list. The cursor in the new list is undefined, regardless of the
// state of the cursor in this List. This List is unchanged.
   List copy(){
      List L = new List();
      Node N = this.front;

      while( N!=null ){
         L.append(N.data);
         N = N.next;
      }
      return L;
   }
*/
}
